components/slack/slack.app.js

```diff
@@ -4,10 +4,102 @@ module.exports = {
   type: "app",
   app: "slack",
   propDefinitions: {
-    conversation: {
+    publicChannel: {
+      type: "string",
+      label: "Channel",
+      //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["public_channel"]
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            if (c.is_im) {
+              return { label: `Direct messaging with: @${userNames[c.user]}`, value: c.id }
+            } else if (c.is_mpim) {
+              return { label: c.purpose.value, value: c.id }
+            } else {
+              return { label: `${c.name}`, value: c.id }
+            }
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    privateChannel: {
       type: "string",
       label: "Channel",
       //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["private_channel"]
+        }
+        const resp = await this.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            return { label: `${c.name}`, value: c.id }
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    user: {
+      type: "string",
+      label: "User",
+      //description: "Select one or more channels or DM conversations to monitor for new messages.",
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          const scopes = await this.scopes()
+          types = ["im"]
+          // TODO use paging
+          userNames = {}
+          for (const user of (await this.users()).users) {
```

Comment:
If you are wondering why your async options are taking so long, it's stuff like this where we are asking for every single user in a workspace before we start to page.  And we do this every time we ask for a new page.

Reply:
Actually it seems like this users call is redundant since this.avaliableConversations is already doing paging calls through the users?